File: Fortran/MLE.f
!! Rutina que usa minuit para realizar Maximum Likelihood que ajusta parámetros de clima. Requiere \emph{f77} para compilar (reemplazada por rutina en C++ usando ROOT).
!!
c	  	to compile and link: f77 -o minfit minfit.f -L /cern/2005/lib -l packlib
c  	where in this case the cern libraries are in /cern/2005/lib
c		Performs a maximum likelihood fit of the model of weather effects given by Lfunc
c		using the minfit procedure from the cern libraries
		implicit double precision(a-h,o-z)
		integer j,i,iutc
c		real*8 logfact
		parameter(nbin = 82536)		!numero de bines desde 01/01/2004 00:00:01 UTC (1072915201) hasta 06/01/2013 00:00:00 UTC (1370044800) 
		dimension ievents(nbin),iutc(nbin)
		dimension pres(nbin),rho(nbin),rhod(nbin),hex6T5(nbin)
		dimension ai(4),error(4),vstrt(4),iflag(4)
		character*10 name(4)
c		dimension xpt(40),ypt(40)
		external Lfunc

		common/likefunc/pres,rho,rhod,hex6T5,ievents,iutc
		common/ref/rho0,P0,pf
c		open(unit=1,file='HeraldData060weatherdrno.dat')
		open(unit=1,file='HeraldData060weatherno.dat')
c		open(unit=1,file='HeraldData060weathergt2004.dat')
c		open(unit=7,file='HeraldData060weatherwinter.dat')
		open(unit=2,file='MLFitout.out')
		open(unit=3,file='mincommands')
		open(unit=12,file='fitmod.dat')
c		open(unit=13,file='fitmodwinter.dat')
c		open(unit=40,file='tmp.dat')
c		open(unit=14,file='fitcoefsdr.dat')
c		open(unit=15,file='HeraldDatasec1weatherdrno.dat')
c		open(unit=15,file='HeraldDatasec1weatherno.dat')
c		open(unit=16,file='HeraldDatasec2weatherdrno.dat')
c		open(unit=16,file='HeraldDatasec2weatherno.dat')
c		open(unit=17,file='HeraldDatasec3weatherdrno.dat')
c		open(unit=17,file='HeraldDatasec3weatherno.dat')
c		open(unit=18,file='HeraldDatasec4weatherdrno.dat')
c		open(unit=18,file='HeraldDatasec4weatherno.dat')
c		open(unit=19,file='HeraldDatasec5weatherdrno.dat')
c		open(unit=19,file='HeraldDatasec5weatherno.dat')

		gamma = 3.23
		B = 1.02
		pf = B*(gamma-1.)
		rho0 = 1.055!1.06d0		   !reference air density in kg/m^3
		P0 = 861.89!862.0d0			!reference pressure in hPa
		call zeroarrs		            !initialize arrays with zeros
		data iflag/4*0/
		data vstrt/ 0.0d0 , 0.0d0 , 0.0d0 , 0.0d0 /

		i = 1	
 11	read (1,*,end=12)iutc(i),ievents(i),pres(i),rho(i),
     &		rhod(i),hex6T5(i)
		i = i + 1
		goto 11
 12	continue

		call mninit(3,2,8)  !  in  out saving
		call fitf(ai,error,name,vstrt,iflag)
		
c		call mnintr(Lfunc,0)
c		goto 10

c		do i=1,4
c			write(6,*) name(i),ai(i),error(i)
c		enddo

		do i=1,3
			write(6,*) name(i),ai(i)/pf,error(i)/pf
		enddo
		write(6,*) name(4),ai(4),error(4)

		call Lfunc2(ai,chi2,ndof,fval,12,smu,isumn)
		write(6,*) '\'chi2\'=',chi2,' \'deg of freed\'=',ndof
     & ,' \'chi2/dof\'=',chi2/float(ndof-4),' \'2lnL\'=',fval

c		call fitunit(7,6,ai,iflag)
		goto 10  !salta al final

		write(3,*)'FIX 4' 
		rewind(unit=3)
		iflag(4) = 1
c	cause Minuit to read commands from the unit IRD originally specified by the user
c	in his call to MNINIT
		call mnintr(Lfunc,0)
		call fitunit(15,14,ai,iflag)
		call fitunit(16,14,ai,iflag)
		call fitunit(17,14,ai,iflag)
		call fitunit(18,14,ai,iflag)
		call fitunit(19,14,ai,iflag)

 10	end

		subroutine fitf(ai,error,name,vstrt,iflag)
		implicit double precision(a-h,o-z)
		dimension ai(4)   ! parameters to be fitted
		dimension arglis(10)          !for Minuit
		dimension error(4), vstrt(4),stp(4),nprm(4)
		dimension iflag(4)
		character*10 name(4)
		external Lfunc

		name(1)='aP';name(2)='arho';name(3)='brho'
		name(4)='lambda'
		data nprm /   1 ,  2   ,  3   ,   4   /
		data stp  / 0.1d0 , 0.1d0  ,  0.1d0 ,  0.1d0  /

c  starting values: id, name, strt, strterr, ? ? ?)
		if(iflag(1).eq.0) then
			call mnparm(nprm(1),name(1),vstrt(1),stp(1),0.0d0,0.0d0,ier)
		endif
		if(iflag(2).eq.0) then
			call mnparm(nprm(2),name(2),vstrt(2),stp(2),0.0d0,0.0d0,ier)
		endif
		if(iflag(3).eq.0) then
			call mnparm(nprm(3),name(3),vstrt(3),stp(3),0.0d0,0.0d0,ier)
		endif
		if(iflag(4).eq.0) then
			call mnparm(nprm(4),name(4),vstrt(4),stp(4),0.0d0,0.0d0,ier)
		endif

c		arglis(2) = 4.
c		arglis(3) = 5.
		call mnexcm(Lfunc,'MIGRAD',arglis,0,ier,0)
c		call mnexcm(Lfunc,'MINOS',arglis,0,ier,0)
c		arglis(1) = 16.
c		call mnexcm(Lfunc,'SET ERR',arglis,1,ier,0)
c		arglis(1) = 1.
c		arglis(2) = 2.
c		arglis(3) = 4.
c		call mnexcm(Lfunc,'MNCONTOUR',arglis,2,ier,0)
		call mnpout(1,name(1),ai(1),error(1),bnd1,bnd2,ier)
		call mnpout(2,name(2),ai(2),error(2),bnd1,bnd2,ier)
		call mnpout(3,name(3),ai(3),error(3),bnd1,bnd2,ier)
		call mnpout(4,name(4),ai(4),error(4),bnd1,bnd2,ier)
c		call mnstat(fmin,fedm,errdef,npari,nparx,istat)
		return
		end

		subroutine Lfunc(npar,grad,fval,ai,iflag,futil)
		implicit double precision(a-h,o-z)
		parameter(nbin = 82536)
		dimension ievents(nbin),iutc(nbin)
		dimension pres(nbin),rho(nbin),rhod(nbin),hex6T5(nbin)
		real*8 mu,logfact
		dimension ai(*),grad(*)
		common/likefunc/pres,rho,rhod,hex6T5,ievents,iutc
		common/ref/rho0,P0,pf
		fval = 0.
		isumn = 0
		sumhex = 0.
		t0=(1220227200-1104537601)/(365.25*24*3600)
c	calculate R0: average rate we would have observed if the atmospheric 
c	parameters were always the reference ones
		do j=1, nbin
c			if(iutc(j).gt.1370044800) exit
			isumn = isumn + ievents(j)
			C=1.+ai(1)*(pres(j)-P0)+ai(2)*(rhod(j)-rho0)+ai(3)*
     &			(rho(j)-rhod(j))
			ti=(iutc(j)-1104537601)/(365.25*24*3600)
			ee=dexp(-ai(4)*(ti-t0))
			if(ee.gt.1.d0) ee = 1.
			sumhex = sumhex + hex6T5(j)*C*ee
c			sumhex = sumhex + hex6T5(j)*C
		end do
		R0 = isumn/sumhex
c	calculate the loglikelihood function
		do j=1, nbin
c			if(iutc(j).gt.1370044800) exit
			C=1.+ai(1)*(pres(j)-P0)+ai(2)*(rhod(j)-rho0)+ai(3)*
     &			(rho(j)-rhod(j))
			ti=(iutc(j)-1104537601)/(365.25*24*3600)
			ee=dexp(-ai(4)*(ti-t0))
			if(ee.gt.1.d0) ee = 1.
			mu = R0*hex6T5(j)*C*ee
c			mu = R0*hex6T5(j)*C
c			if(mu.ge.3.d0.and.ievents(j).eq.0) goto 20
c			if(mu.lt.0.5d0) goto 20
			if(mu.eq.0.0d0) goto 20
			fval = fval-2*( ievents(j)*dlog(mu)-mu-logfact(ievents(j)) )
 20	continue
		end do
		return	
		end		

c	function to calculate the natural logarithm to the factorial of N
		real*8 function logfact(N)
			integer N,I
			logfact=0.
			if(N.gt.1)then
			 	do I=2,N 
			 		logfact = logfact + dlog(dfloat(I)) 					
			 	end do
			end if 
		return 
		end


		subroutine Lfunc2(ai,chi2,ndof,fval,iunit,smu,isumn)
		implicit double precision(a-h,o-z)
		parameter(nbin = 82536)
		dimension ievents(nbin),iutc(nbin)
		dimension pres(nbin),rho(nbin),rhod(nbin),hex6T5(nbin)
		real*8 mu,logfact
		dimension ai(*)
		common/likefunc/pres,rho,rhod,hex6T5,ievents,iutc
		common/ref/rho0,P0,pf
		fval = 0.
		isumn = 0
		sumhex = 0.
		chi2 = 0.
		ndof = 0
		smu = 0.
		t0=(1220227200-1104537601)/(365.25*24*3600)
c	calculate R0: average rate we would have observed if the atmospheric 
c	parameters were always the reference ones
		do j=1, nbin
c			if(iutc(j).gt.1370044800) exit
			isumn = isumn + ievents(j)
			C=1.+ai(1)*(pres(j)-P0)+ai(2)*(rhod(j)-rho0)+ai(3)*
     &			(rho(j)-rhod(j))
			ti=(iutc(j)-1104537601)/(365.25*24*3600)
			ee=dexp(-ai(4)*(ti-t0))
			if(ee.gt.1.d0) ee = 1.
			sumhex = sumhex + hex6T5(j)*C*ee
c			sumhex = sumhex + hex6T5(j)*C
		end do
		R0 = isumn/sumhex
c	calculate the loglikelihood function
		do j=1, nbin
c			if(iutc(j).gt.1370044800) exit
			C=1.+ai(1)*(pres(j)-P0)+ai(2)*(rhod(j)-rho0)+ai(3)*
     &			(rho(j)-rhod(j))
			ti=(iutc(j)-1104537601)/(365.25*24*3600)
			ee=dexp(-ai(4)*(ti-t0))
			if(ee.gt.1.d0) ee = 1.
			mu = R0*hex6T5(j)*C*ee
c			mu = R0*hex6T5(j)*C
c			if(mu.ge.3.d0.and.ievents(j).eq.0) goto 20
c			if(mu.lt.0.5d0) goto 20
			if(mu.eq.0.0d0) goto 20
			smu = smu + mu
			chi2 = chi2 + (dfloat(ievents(j))-mu)**2/mu
			ndof = ndof + 1
			fval = fval+2*( ievents(j)*dlog(mu)-mu-logfact(ievents(j)) )
			write(iunit,*)iutc(j),mu,ievents(j),sqrt(float(ievents(j)))
     &			,hex6T5(j),ti*365.25
 20		continue
		end do
		return	
		end

c     fills the common arrays with zeroes
		subroutine zeroarrs
		implicit double precision(a-h,o-z)
		parameter(nbin = 82536)
		dimension ievents(nbin),iutc(nbin)
		dimension pres(nbin),rho(nbin),rhod(nbin),hex6T5(nbin)
		common/likefunc/pres,rho,rhod,hex6T5,ievents,iutc
		do i=1,nbin
			ievents(i)=0
			iutc(i)=0
			pres(i)=0.
			rho(i)=0.
			rhod(i)=0.
			hex6T5(i)=0.
		enddo
		return
		end

c     runs the fit procedure taking the data from file in unit 'iunitin'
c     and the previous fit parameter and writes the new fit parameters in 'iunitout'
		subroutine fitunit(iunitin,iunitout,vstrt,iflag)
		implicit double precision(a-h,o-z)
		parameter(nbin = 82536)
		dimension ievents(nbin),iutc(nbin)
		dimension pres(nbin),rho(nbin),rhod(nbin),hex6T5(nbin)
		common/likefunc/pres,rho,rhod,hex6T5,ievents,iutc
		common/ref/rho0,P0,pf
		dimension ai(4),error(4),vstrt(4),iflag(4)
		character*10 name(4)
		integer status

		call zeroarrs

		i = 1
 11	read (iunitin,*,end=12)iutc(i),ievents(i),pres(i),rho(i),
     &		rhod(i),hex6T5(i)
		i = i + 1
		goto 11
 12	continue

		call fitf(ai,error,name,vstrt,iflag)
		write(iunitout,*) ''
c		do i=1,4
c			write(iunitout,*) name(i),ai(i),error(i)
c		end do

		do i=1,3
			write(iunitout,*) name(i),ai(i)/pf,error(i)/pf
		enddo
		write(iunitout,*) name(4),ai(4),error(4)

		call Lfunc2(ai,chi2,ndof,fval,13,smu,isumn)
		write(iunitout,*) '\'chi2\'=',chi2,' \'deg of freed\'=',ndof
     & ,' \'chi2/dof\'=',chi2/float(ndof-3),' \'2lnL\'=',fval
		
		do k=1,5;vstrt(k)=ai(k);enddo
		return
		end

File: Sh/MLE.c


#include "TMinuit.h"
#include <stdlib.h>
#include <stdio.h>
#include <math.h>

const Double_t rho0 = 1.05236;
const Double_t P0 =861.729;

////////////////////////////////////////////
const int nbins = 71451;
const int initial_time= 1104550200;
const char* filename= "/home/ponci/Desktop/TesisIB/Coronel/Merged_Herald_Weather/Energy_above_1EeV/Old/Herald_old_sector_2_weather.dat"; 
////////////////////////////////////////////

const char* output_file="/home/ponci/Desktop/TesisIB/Coronel/Merged_Herald_Weather/fitted_parameters.dat";


//Arrays with the data
Float_t pres[nbins],rho[nbins],rhod[nbins],hex6T5[nbins], rate[nbins];
Int_t ievents[nbins],iutc[nbins];

//________________________________________________________________________
///	function to calculate the natural logarithm of the factorial of N
Double_t logfact(Int_t N){
	int i;
	Double_t logf = 0.;
	if(N > 1){
		for (i = 2; i <= N; i++) logf += log((double) i); 
		}
		return logf; 	}


//______________________________________________________________________________
/// input function for minuit, structure required by Tminuit in order to minimize
void fcn(Int_t &npar, Double_t *gin, Double_t &f, Double_t *par, Int_t iflag){

//calculate log Likelihood
   Double_t flogL = 0.;
   Int_t isumn = 0;
   Double_t sumhex = 0.;
   Double_t ti,t0,C,R0,mu,ee;
  
   t0=(nbins*3600)/(365.25*24*3600);

//	calculate R0: average rate we would have observed if the atmospheric parameters were always the reference ones
	for (int j = 0; j < nbins; j++) 
		{
			isumn += ievents[j];
			C=1.+par[0]*(pres[j]-P0)+par[1]*(rhod[j]-rho0)+par[2]*(rho[j]-rhod[j]);

			sumhex += hex6T5[j]*C;
		}

		R0 = isumn/sumhex;

//Maximum Likehood Estimator
	for (int j = 0; j < nbins; j++) 
		{
			C=1.+par[0]*(pres[j]-P0)+par[1]*(rhod[j]-rho0)+par[2]*(rho[j]-rhod[j]);

			mu = R0*hex6T5[j]*C;
			if(mu == 0.0) continue;
			flogL -= 2*( ievents[j]*log(mu)-mu-logfact(ievents[j]) );
	   }

//F must be evaluated before exiting the FCN function
   f = flogL;
}

//______________________________________________________________________________
void fcn2(Double_t *par,Double_t *pchi2,Int_t *pndof)
{
//calculate log Likelihood
	int j;
	Double_t flogL = 0.;
	Int_t isumn = 0;
	Double_t sumhex = 0.;
	Double_t ti,t0,C,R0,mu,ee,chi2;
	Int_t ndof;
	chi2 = 0.;
	ndof = 0;
	t0=(nbins*3600)/(365.25*24*3600);

//	calculate R0: average rate we would have observed if the atmospheric 
//	parameters were always the reference ones
	for (j = 0; j < nbins; j++) {
			isumn += ievents[j];
			C=1.+par[0]*(pres[j]-P0)+par[1]*(rhod[j]-rho0)+par[2]*(rho[j]-rhod[j]);
			sumhex += hex6T5[j]*C;
	}
		R0 = isumn/sumhex;
	printf("\n EL RATE : %f", R0);

	for (j = 0; j < nbins; j++) {
			C=1.+par[0]*(pres[j]-P0)+par[1]*(rhod[j]-rho0)+par[2]*(rho[j]-rhod[j]);

			mu = R0*hex6T5[j]*C;

			if(mu == 0.0) continue;
			flogL -= 2*( ievents[j]*log(mu)-mu-logfact(ievents[j]) );
			ndof++;
			chi2 += ((double)ievents[j]-mu)*((double)ievents[j]-mu)/mu;
   }
   *pchi2 = chi2;
   *pndof = ndof;
}

//______________________________________________________________________________
void MLE()
{	
// data input  
	FILE *in_data = fopen(filename,"r"); 

	if (! in_data ) // equivalent to saying if ( in_data == NULL ) 
    	{  
		printf("Error! File can't be read\n"); 
        	exit(-1); 
    	}	
//Read from datafile
    for (int i=0; i < nbins; i++) {
    	fscanf(in_data, "%d %d %g %g %g %g %g",&iutc[i], &ievents[i],&pres[i],&rho[i], &rhod[i], &hex6T5[i]), &rate[i];}
	
	TMinuit *gMinuit = new TMinuit(4);  //initialize TMinuit with a maximum of 4 params
	gMinuit->SetFCN(fcn);				//Set the function to be minimized

	Double_t arglist[10];
	Int_t ierflg = 0;

	arglist[0] = 1;
	gMinuit->mnexcm("SET ERR", arglist ,1,ierflg);

// Set starting values and step sizes for parameters

	printf("\n\n\n\nSet starting values and step sizes for parameters\n" );
	
	static Double_t vstart[4] 	= {0.0 , 0.0 , 0.0 , 0.0};
	static Double_t step[4] 	= {0.1 , 0.1 , 0.1 , 0.1};	

	gMinuit->mnparm(0, "aP", 		vstart[0], step[0], 0,	0,	ierflg);
	gMinuit->mnparm(1, "arho", 		vstart[1], step[1], 0,	0,	ierflg);
	gMinuit->mnparm(2, "brho", 		vstart[2], step[2], 0,	0,	ierflg);
	gMinuit->mnparm(3, "Lambda",	vstart[3], step[3], 0,	0,	ierflg);

// Now ready for minimization step

	printf("\n\n\n\nNow ready for minimization step\n" );
	arglist[0] = 500;
	arglist[1] = 1.;
	gMinuit->mnexcm("MIGRAD", arglist ,2,ierflg);
	
	Double_t amin,edm,errdef;
	Int_t nvpar,nparx,icstat;
	gMinuit->mnstat(amin,edm,errdef,nvpar,nparx,icstat);
   	gMinuit->mnprin(3,amin);

   	in_data = fopen(filename,"r"); 
	if (! in_data ) // equivalent to saying if ( in_data == NULL ) 
    	{  
		printf("Error! File can't be read\n"); 
        	exit(-1); 
    	}	
	for (int i=0; i < nbins; i++){
		fscanf(in_data, "%d %d %g %g %g %g",&iutc[i], &ievents[i],&pres[i],&rho[i],
			&rhod[i],&hex6T5[i]);
	}

	arglist[0] = 4;
	gMinuit->mnexcm("FIX", arglist ,1,ierflg);
	arglist[0] = 500;
	arglist[1] = 1.;
	gMinuit->mnexcm("MIGRAD", arglist ,2,ierflg);

// Get parameter results

	Double_t pars[4],errors[4],bnd1,bnd2,chi2,fval;
	Int_t ndof,ivar;
	char* pname;
	TString parn(pname);
	gMinuit->mnpout(0, parn, pars[0], errors[0], bnd1,	bnd2,	ivar);
	gMinuit->mnpout(1, parn, pars[1], errors[1], bnd1,	bnd2,	ivar);
	gMinuit->mnpout(2, parn, pars[2], errors[2], bnd1,	bnd2,	ivar);
	gMinuit->mnpout(3, parn, pars[3], errors[3], bnd1,	bnd2,	ivar);


// Print results         
	fcn2(pars,&chi2,&ndof);

	Double_t gamma = 3.23;
	Double_t B = 1.02;
	Double_t pf = B*(gamma-1.);

	pf=1;

	printf("\naP	= %g +/- %g \n",pars[0]/pf,	errors[0]/pf);
	printf("arho	= %g +/- %g \n",pars[1]/pf,	errors[1]/pf);
	printf("brho	= %g +/- %g \n",pars[2]/pf,	errors[2]/pf);
	printf("Lambda	= %g +/- %g \n",pars[3],	errors[3]);

	printf("chi2	= %f  ndof= %d  chi2/ndof= %f\n",chi2,ndof,chi2/ndof);

		FILE *out_data= fopen(output_file, "a+");

	fprintf(out_data, "Data file: %s 	\n", filename);
	fprintf(out_data, "aP		= %g +/- %g \n",pars[0]/pf,	errors[0]/pf);
	fprintf(out_data, "arho		= %g +/- %g \n",pars[1]/pf,	errors[1]/pf);
	fprintf(out_data, "brho		= %g +/- %g \n",pars[2]/pf,	errors[2]/pf);

	printf("chi2	= %f  ndof= %d  chi2/ndof= %f\n",chi2,ndof,chi2/ndof);

	fclose(out_data);

}

//______________________________________________________________________________
#ifndef __CINT__
int main(int argc, char** argv) {

MLE();
return 0;
}
#endif
